Libraries/LibPthread/pthread.cpp

```diff
@@ -157,70 +167,233 @@ int pthread_sigmask(int how, const sigset_t* set, sigset_t* old_set)
     return 0;
 }
 
-int pthread_mutex_init(pthread_mutex_t* mutex, const pthread_mutexattr_t* attributes)
+#define PTHREAD_FUTEX_UNLOCKED 0u
+#define PTHREAD_FUTEX_LOCKED 1u
+#define PTHREAD_FUTEX_LOCKED_WITH_WAITERS 2u
+
+static int futex_lock(uint32_t& futex_addr, uint32_t futex_flags, const timespec* abstime = nullptr, bool no_wait = false)
```

Comment:
I don't think it's a good idea to wrap futexes like that to make them appear to be simple locks. Futexes are much more versatile: one can, and should, use the value of the futex word directly to encode the current state (such as "locked for writing, no waiting readers"), not just "locked" and "unlocked".

Reply:
Maybe I'll call the function `mutex_lock` then, it's an internal function only...

Reply:
My point is rather that each synchronization primitive should use futexes in its own special way that makes sense for that primitive — not simply as a lock.

I especially dislike it when various operations get implemented by locking an internal lock, checking some state, then unlocking the internal lock. That kills performance. Operations should not block each other from progressing any more than is required by the semantic of the operation — i.e., if we're talking about `Mutex::lock()` then surely it must block other threads executing `Mutex::lock()`, but `ConditionVariable::notify_all()` should not block other threads from doing the same. Also, "acquire" operations should not have to use any release-or-stronger atomics, and vice versa.

Reply:
Yeah i get your point, but in the case of condition variables I can't figure out a way to do it without essentially a mutex. Even glibc uses a mutex for waiting, signaling, and broadcasting a condition variable. Every single time.

That being said, I would really like to leverage `FUTEX_WAIT_OP` and `FUTEX_CMP_REQUEUE` but I haven't been able to come up with a scheme that would work. Any ideas?

Reply:
I've posted a link to my implementation of a condition variable (and other synchronization primitives, too!) somewhere above. [Here it is again](https://github.com/bugaevc/lets-write-sync-primitives/blob/master/src/condvar.cpp). It's supposed to be understandable, clean code, and it uses `FUTEX_REQUEUE` 😀 (without `CMP` though).

Reply:
Yeah I remember that one. This implementation also uses `FUTEX_REQUEUE`. Your implementation doesn't allow condvars choosing an arbitrary mutex after creation, which means you can't create a condvar and pick a mutex later (or change it). While condvars only allow the same mutex to be used at the same time, it's entirely possible to use a different mutex with the same condvar at a later point, so long as it's consistent. Hence the "mutex" that e.g. glibc uses in their condvar implementation.

Reply:
Yes, I took the opportunity to pick somewhat nicer APIs there 🙂

TBH using a single condvar with multiple mutexes is a weird/broken thing to do to begin with (unlike using a single mutex with multiple condvars). But if we have to support that, it should be a matter of just doing relaxed reads/writes to a mutex pointer, no?

Reply:
Quite frankly, it'd be a nice LibCore thing ;) It'd be nice to have c++ mutexes, condvars etc, and they do not have to be built upon pthreads necessarily ;-)

Reply:
That's what LibThread is for 🙂

But the issue with that is we want both LibThread to use LibPthread and the other way around, and that doesn't quite work.

Reply:
No, it's not broken. Yeah the API sucks, and if the mutex had to be specified when initializing the condvar then this particular model could work. The standard only says that you cannot use different mutexes at the same time, so you can start out with mutex A, and when no thread is waiting switch to using mutex B using the same condvar. But we only need pthread to support ports, we can have a better, more restrictive but also more performant condvar like what you came up with for our own stuff in LibCore.

Reply:
> No, it's not broken.

Is there an example of how a condvar can be reasonably used with multiple mutexes? In my mental model, a condvar is supposed to guard a single condition/predicte that's also protected by a mutex (you can only evaluate or modify the condition while holding the mutex). So I can't really imagine a valid use case for different `wait()` calls supplying different mutexes.

> The standard only says that you cannot use different mutexes at the same time, so you can start out with mutex A, and when no thread is waiting switch to using mutex B using the same condvar.

Yes, I understand. I'm saying I don't understand why that would be useful for a program to use a condvar this way, because to me it sounds like breaking the ideas condvars are built on.

> Yeah the API sucks, and if the mutex had to be specified when initializing the condvar then this particular model could work. <...> we can have a better, more restrictive but also more performant condvar like what you came up with for our own stuff in LibCore.

But again, why can't my implementation be made to work with changing mutexes too? You'd replace the reference to the mutex with a pointer, set the mutex first thing in `.wait()`, and read it in `.notify_all()`. Both reading and writing the mutex pointer ~can be relaxed~ don't even have to be atomic (because the mutex pointer gets essentially protected by the mutex itself); no need for additional locking.

Reply:
I'm not sure how you can get this without races without a mutex even if you use an atomic variable for the mutex pointer. Think of two rapid broadcasts/notify_all where the waiters switch the mutex (while assuring they aren't using a different mutex at the same time). Depending on the sequence of operations in the wait that changes the mutex and the second broadcast/notify_all, you could end up using the previous mutex accidentally. The only way I can think of that you can do this is using a mutex. This is the line I'm referring to in this implementation:
https://github.com/SerenityOS/serenity/blob/3ff4d2349a416c011fee8f50ea048314834c2475/Libraries/LibPthread/pthread.cpp#L827

As you can see, it gets the mutex pointer and then resets it. Because this is synchronized with wait/signal/broadcast it should be race-free, even though it is using the pointer obtained while holding the lock after releasing the lock. This should still be still race-free as a subsequent wait with a different mutex between grabbing the pointer and actually waking/requeueing would just requeue to the previous mutex, which should still exist and would be the correct mutex to requeue to based on when broadcast was triggered. Any subsequent signal/broadcast would requeue to the new mutex, even if the first requeue hasn't happened yet.

Reply:
Oh, I think I see what you mean.

That (using a single condvar with different mutexes *concurrently*) is *even more* broken than "just" using a single condvar with different mutexes. Surely we don't have to support that!

> would just requeue to the previous mutex, which should still exist and would be the correct mutex to requeue to based on when broadcast was triggered

I'm not sure there's such a thing as a thing as a "correct" mutex to use here, because this whole situation is very wrong. There can be no happens-before relationship between entering `wait()` and `notify()` calls (even if some external means of synchronization are used). And if there was, there's no guarantee that they'll acquire you internal lock in the same order.

I would argue that the only "correct" thing to do is for the wake/notify path to requeue waiting threads onto whichever mutex the condition variable is bound to "at that very moment", or in other words that fetching the mutex and requeuing must be done atomically wrt `wait()` calls changing the mutex. If that is done atomically and the `wait()` calls are somehow synchronized between themselves, all the waiters at the moment of requeuing use the same mutex — the one we'd atomically observe — and we'd wake/requeue them all "correctly".

That could be factored into you implementation, you just have to release the internal lock later after requeuing. That would slow it down further though, since you'll hold onto the lock for the duration of the syscall.

The other "correct" thing to do is to not use requeuing at all (i.e. pessimize) in case concurrent mutex modification is detected. Or, since we can only detect this *after* requeuing, wake everyone we've requeued up, effectively undoing the requeuing. My implementation could be patched the following way for that:

```cpp
void CondVar::notify_all() {
    // ...increment state...
    if (need to wake) {
        // ...unset need to wake bits...
        Mutex *mutex1 = m_mutex.load();
        futex_requeue(
            (const uint32_t *) &state, 1,
            (const uint32_t *) &mutex1->state, INT_MAX
        );
        Mutex *mutex2 = m_mutex.load();
        if (mutex1 != mutex2) {
            // Mutex has changed! Make sure
            // to wake everyone we may just have
            // requeued to the possibly wrong mutex,
            // and let them sort it out themselves.
            futex_wake(&mutex1->state, INT_MAX);
        }
    }
}
```

But again, IMO this is too broken to support. We don't have to follow POSIX 100%.

Reply:
> `mutex1 != mutex2`

No, can't do that, that's the ABA problem.

Reply:
Well, I guess if we can't figure that out *and* we decide we have to support this case — it'd be fine to use an internal lock on the slow path. For condvars, `wait()` is always the slow path, and `notify_all()`/`broadcast()` has a fast path where no one's waiting and a slow path where we do have to wake somebody; it would be OK to use the lock on the latter. We should make sure to drop the "have waiters"/"should take slow path" flag prior to taking the lock though, so that of two concurrent `notify_all()` calls only one has to enter the slow path.

Reply:
I don't think the re-queueing has to be done while holding the lock. But yeah there is a race where changing the mutex could cause requeuing to the wrong mutex. I wonder if this could be solved somehow with `FUTEX_CMP_REQUEUE` and maybe yet another atomic variable tracking mutex changes or something like that... Maybe I'll just drop the usermode changes from this PR for now, because I think the kernel implementation is conceptually matching that of Linux. I haven't tested it out thoroughly, though...

Reply:
> I wonder if this could be solved somehow with `FUTEX_CMP_REQUEUE`

I've thought about that, but that wouldn't work because it compares the value of the same word it requeues from, not the value of a different word.

We could add a special kernel op for this; but I'm not sure it'd be much better than locking in userspace on the slow path.